core/src/main/scala/org/apache/spark/CacheManager.scala

```diff
@@ -46,7 +48,14 @@ private[spark] class CacheManager(blockManager: BlockManager) extends Logging {
         val existingMetrics = context.taskMetrics().registerInputMetrics(blockResult.readMethod)
         existingMetrics.incBytesReadInternal(blockResult.bytes)
 
-        val iter = blockResult.data.asInstanceOf[Iterator[T]]
+        val iter = {
+          val dataIter = blockResult.data.asInstanceOf[Iterator[T]]
+          if (blockResult.readMethod != DataReadMethod.Network) {
+            CompletionIterator[T, Iterator[T]](dataIter, blockManager.releaseLock(key))
```

Comment:
this warrants a comment; why do we treat reading network blocks differently? (that the lower level network layer already releases the locks so we don't want to do it here again)